server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,290 @@
+    }
+
+    public boolean isTopicIdsSupported() {
+        return this.isAtLeast(IBP_2_8_IV0);
+    }
+
+    public boolean isAllocateProducerIdsSupported() {
+        return this.isAtLeast(IBP_3_0_IV0);
+    }
+
+
+    public RecordVersion highestSupportedRecordVersion() {
+        if (this.isLessThan(IBP_0_10_0_IV0)) {
+            return RecordVersion.V0;
+        } else if (this.isLessThan(IBP_0_11_0_IV0)) {
+            return RecordVersion.V1;
+        } else {
+            return RecordVersion.V2;
+        }
+    }
+
+    private static final Map<String, MetadataVersion> IBP_VERSIONS;
+    static {
+        {
+            IBP_VERSIONS = new HashMap<>();
+            Map<String, MetadataVersion> maxInterVersion = new HashMap<>();
+            for (MetadataVersion metadataVersion : MetadataVersion.values()) {
+                maxInterVersion.put(metadataVersion.apacheRelease, metadataVersion);
+                IBP_VERSIONS.put(metadataVersion.name, metadataVersion);
+            }
+            IBP_VERSIONS.putAll(maxInterVersion);
+        }
+    }
+
+    public String shortVersion() {
+        return apacheRelease;
+    }
+
+    public String version() {
+        return name;
+    }
+
+    /**
+     * Return an `MetadataVersion` instance for `versionString`, which can be in a variety of formats (e.g. "0.8.0", "0.8.0.x",
+     * "0.10.0", "0.10.0-IV1"). `IllegalArgumentException` is thrown if `versionString` cannot be mapped to an `MetadataVersion`.
+     * Note that 'misconfigured' values such as "1.0.1" will be parsed to `IBP_1_0_IV0` as we ignore anything after the first
+     * two digits for versions that don't start with "0."
+     */
+    public static MetadataVersion fromVersionString(String versionString) {
+        String[] versionSegments = versionString.split(Pattern.quote("."));
+        int numSegments = (versionString.startsWith("0.")) ? 3 : 2;
+        String key;
+        if (numSegments >= versionSegments.length) {
+            key = versionString;
+        } else {
+            key = String.join(".", Arrays.copyOfRange(versionSegments, 0, numSegments));
+        }
+        return Optional.ofNullable(IBP_VERSIONS.get(key)).orElseThrow(() ->
+            new IllegalArgumentException("Version " + versionString + " is not a valid version")
+        );
+    }
+
+    /**
+     * Return the minimum `MetadataVersion` that supports `RecordVersion`.
+     */
+    public static MetadataVersion minSupportedFor(RecordVersion recordVersion) {
+        switch (recordVersion) {
+            case V0:
+                return IBP_0_8_0;
+            case V1:
+                return IBP_0_10_0_IV0;
+            case V2:
+                return IBP_0_11_0_IV0;
+            default:
+                throw new IllegalArgumentException("Invalid message format version " + recordVersion);
+        }
+    }
+
+    public static MetadataVersion latest() {
+        MetadataVersion[] values = MetadataVersion.values();
```

Comment:
Unfortunately, `values()` causes the array to be allocated each time. So, we typically store the result of `values` in a private static final field and then access the elements from that instead.